src/liballoc/collections/btree/set.rs

```diff
@@ -314,24 +324,46 @@ impl<T: Ord> BTreeSet<T> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn difference<'a>(&'a self, other: &'a BTreeSet<T>) -> Difference<'a, T> {
-        if self.len() > other.len() / ITER_PERFORMANCE_TIPPING_SIZE_DIFF {
-            // Self is bigger than or not much smaller than other set.
-            // Iterate both sets jointly, spotting matches along the way.
-            Difference {
-                inner: DifferenceInner::Stitch {
-                    self_iter: self.iter(),
-                    other_iter: other.iter().peekable(),
-                },
-            }
-        } else {
-            // Self is much smaller than other set, or both sets are empty.
-            // Iterate the small set, searching for matches in the large set.
-            Difference {
-                inner: DifferenceInner::Search {
+        Difference {
+            inner: if self.len() <= 1 || other.len() <= 1 {
+                // Optimizations below don't work or don't pay off.
+                DifferenceInner::Search {
                     self_iter: self.iter(),
                     other_set: other,
-                },
-            }
+                }
+            } else if self.len() <= 4 && other.len() <= 4 {
+                // Optimizations below don't pay off.
+                DifferenceInner::Stitch {
+                    self_iter: self.iter(),
+                    other_iter: other.iter().peekable(),
+                }
+            } else {
+                let mut self_except_first = self.iter();
+                let self_min = self_except_first.next().unwrap();
+                let mut self_except_last = self.iter();
+                let self_max = self_except_last.next_back().unwrap();
+                let mut other_iter = other.iter();
+                let other_min = other_iter.next().unwrap();
+                let other_max = other_iter.last().unwrap();
```

Comment:
next_back and last are used interchangeably. I guess `last` is preferred for readability(?) but `next_back` has its own readability benefit — we don't have to check back that the iterator has special case `last`.

Reply:
`last` consumes the iterator, so it can't be used the other places, which try to reuse the iterator (in so far this small optimization is worth our trouble). Thus it seemed more clear to use it whenever possible, to say you don't need the iterator anymore. Come to think of it, a separate function that shields the iterator entirely and returns a pair of first and last might be better.

I don't understand "we don't have to check back that the iterator has special case last". Compiler said yes, so it's there?

Reply:
The default last() will traverse the whole iterator, so at least to review, you'll have to check that this is not the case here. And it's not.

Reply:
Ooh... right. I just spotted it as a sort of alternative to next_back in the iterator.

I'll use next_back for consistency.